### ISSUE TITLE
MacPass writes updated database on disk even if unnecessary

### ISSUE THREAD

**Issue by lifepillar** [mstarke/MacPass#570]
When saving a file that has not been edited, or even when just locking it, MacPass updates the database on disk. Although not strictly a bug, this is unnecessary.

**Comment by benwalch** [comment]
+1

**Comment by mstarke** [comment]
If the user issues a save the file is written regardless of state. You can compare this behaviour e.g. to TextEdit and it works the same way. I would suggest no change to that since it's a use based action.

The second thing is lock. This was different in previous releases of MacPass and resulted in data loss for a lot of users. Currently there's autosave in place so this issue should not arise anymore. My reasoning for this behaviour:

### External changes
If the database is locked and changed in any way, changes should be merged. Merges cannot take place without unlocking the database and asking for merge strategies. 0.6.2 does not have merge support, it only allows to decide for reload or overwrite on the event of an external change to the database. 0.7 will include merge support. By saving on lock, the need for user interaction is removed. Also, this is a good way to propagate updates to other clients if the file cannot be saved frequently via autosave mechanics of NSDocument.

### Master Key Changes
If the user changes the master key this change is not affective until the next save. With autosave this should be immediate but it might be postponed in certain scenarios. By saving on lock the master password change is in effect directly after the lock (which still might never occur)

###  Keepass
Keepass writes on lock, so it seemed appropriate to do the same.

**Comment by lifepillar** [comment]
>If the user issues a save the file is written regardless of state. You can compare this behaviour e.g. to TextEdit and it works the same way.

True. But is it typical behaviour, or one recommended by Apple? Anyway, I haven't checked how TextEdit writes such a document, but MacPass writes a *different* file every time even if no change has been made (I mean, not only its attributes, such as the modified timestamp, are updated, but also its content, although not in an apparent way). I find that a bit counterintuitive and confusing, especially for an app that deals with privacy and security.

>The second thing is lock. This was different in previous releases of MacPass and resulted in data loss for a lot of users.

I am not asking to avoid saving a document altogether when locking, but to avoid it when the document has not been modified at all since it was opened or last saved. I do not see how that could result in data loss: you simply need to close the document's view and wipe memory. It would also make locking much faster (now, there's a slight delay because the document is always saved).

Btw, there is something wrong in the way MacPass (0.6.2) displays a document status in the top-left close button. If I open a document and a new entry, the new entry immediately appears in the main table. The window's close button, however, does not reflect that. If I start typing inside the title field, the close button changes to a modified state, but as soon as I press tab to move to the next field, it switches back to unmodified state. Despite such changes in the UI, the document is never written to disk.

**Comment by lifepillar** [comment]
> MacPass writes a different file every time even if no change has been made (I mean, not only its attributes, such as the modified timestamp, are updated, but also its content, although not in an apparent way)

Sorry, forget about that: it's not true.

**Comment by mstarke** [comment]
0.6.2. has a lot of broken stuff regarding undo/redo and correct handling of the modified status of the database. But that should all be resolved by now.

About the file diffs. This is behaviour as intended. Take a look at http://keepass.info/help/base/security.html#secdictprotect for more information about the rationale.

**Comment by mstarke** [comment]
The current continuous build should only write a file if things did change (based on the edited flag on a file). Can you confirm if this is the case or does MacPass still behaves unexpectedly?

**Comment by eskyto** [comment]
First of all, thank you for a great application. I have just started to use it.

I can confirm that with version 0.7.3 (17830), if I lock the database without changing anything, then the .kdbx file is changed.

**Comment by djkelly99** [comment]
Just wondering if I should create a new issue or if my issue is the same root cause as this one.  If my database file is in google drive and I concurrently use 2 instances of Mac Pass on different machines, a database merge loop will occur whereby without actually making any changes, the Merge -> Save Anyway dialog will bounce from my machine to the other.  It's like the Merge action updates the file, which the other instance detects, which then feels like it has to merge again, and so on, eventhough no changes have actually taken place.  Any suggested workaround to allow 2 instance of mac pass read from the one common database file?

**Comment by georgesnow** [comment]
@djkelly99 are you logged into both Macs at the same time?

**Comment by mstarke** [comment]
@djkelly99 please create a new issue. I think your setup is something that exceeds the built-in functionality of macOS.

**Comment by djkelly99** [comment]
@georgesnow Yes, I'm logged into 2 macs at the same time.  I apprecciate its' asking a lot from a locking/unlocking point of view. Will create a new issue to track. Thanks

**Comment by georgesnow** [comment]
As note I experienced this as well(in my case it was iCloud).This part of the reason I added the “lock on screen sleep”. I move between a desktop and laptop throughout the day and sometimes I would forget to lock a database on one of the machines and this happened.  I can post to your new ticket.

**Comment by georgesnow** [comment]
This mostly resolved the issue. Cause I use the lock screen and about 10 seconds later the displays go to sleep and subsequently lock my database. But if you are logged into  both at the same time that won’t help you.

**Comment by djkelly99** [comment]
thanks, I'll the workaround in mind, it might be of some use now and again.  But like you said, it'll only work if both machines aren't used simultaneously.  created new ticket [here](https://github.com/MacPass/MacPass/issues/1036) to track.
